source/adapters/level_zero/image.cpp

```diff
@@ -36,75 +586,151 @@ urBindlessImagesUnsampledImageHandleDestroyExp(ur_context_handle_t hContext,
+             (hContext->ZeContext, hDevice->ZeDevice, ZeImage));
+  UR_CALL(createUrMemFromZeImage(hContext, ZeImage, /*OwnZeMemHandle*/ true,
+                                 ZeImageDesc, phImageMem));
+  return UR_RESULT_SUCCESS;
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL urBindlessImagesImageFreeExp(
     ur_context_handle_t hContext, ur_device_handle_t hDevice,
     ur_exp_image_mem_handle_t hImageMem) {
   std::ignore = hContext;
   std::ignore = hDevice;
-  std::ignore = hImageMem;
-  urPrint("[UR][L0] %s function not implemented!\n", __FUNCTION__);
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+  UR_CALL(urMemRelease(reinterpret_cast<ur_mem_handle_t>(hImageMem)));
+  return UR_RESULT_SUCCESS;
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL urBindlessImagesUnsampledImageCreateExp(
     ur_context_handle_t hContext, ur_device_handle_t hDevice,
     ur_exp_image_mem_handle_t hImageMem, const ur_image_format_t *pImageFormat,
     const ur_image_desc_t *pImageDesc, ur_mem_handle_t *phMem,
     ur_exp_image_handle_t *phImage) {
-  std::ignore = hContext;
-  std::ignore = hDevice;
-  std::ignore = hImageMem;
-  std::ignore = pImageFormat;
-  std::ignore = pImageDesc;
-  std::ignore = phMem;
-  std::ignore = phImage;
-  urPrint("[UR][L0] %s function not implemented!\n", __FUNCTION__);
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+  std::shared_lock<ur_shared_mutex> Lock(hContext->Mutex);
+
+  ZeStruct<ze_image_desc_t> ZeImageDesc;
+  UR_CALL(ur2zeImageDesc(pImageFormat, pImageDesc, ZeImageDesc));
+
+  ze_image_handle_t ZeImage;
+
+  ze_memory_allocation_properties_t MemAllocProperties{
+      ZE_STRUCTURE_TYPE_MEMORY_ALLOCATION_PROPERTIES};
+  ZE2UR_CALL(zeMemGetAllocProperties,
+             (hContext->ZeContext, hImageMem, &MemAllocProperties, nullptr));
+  if (MemAllocProperties.type == ZE_MEMORY_TYPE_UNKNOWN) {
+    _ur_image *UrImage = reinterpret_cast<_ur_image *>(hImageMem);
+    if (!isSameImageDesc(&UrImage->ZeImageDesc, &ZeImageDesc)) {
+      ze_image_bindless_exp_desc_t ZeImageBindlessDesc;
+      ZeImageBindlessDesc.stype = ZE_STRUCTURE_TYPE_BINDLESS_IMAGE_EXP_DESC;
+      ZeImageBindlessDesc.pNext = nullptr;
+      ZeImageBindlessDesc.flags = ZE_IMAGE_BINDLESS_EXP_FLAG_BINDLESS;
+      ZeImageDesc.pNext = &ZeImageBindlessDesc;
+      ZE2UR_CALL(zeImageViewCreateExt,
+                 (hContext->ZeContext, hDevice->ZeDevice, &ZeImageDesc,
+                  UrImage->ZeImage, &ZeImage));
+      ZE2UR_CALL(zeContextMakeImageResident,
+                 (hContext->ZeContext, hDevice->ZeDevice, ZeImage));
+      UR_CALL(createUrMemFromZeImage(hContext, ZeImage, /*OwnZeMemHandle*/ true,
+                                     ZeImageDesc, phMem));
+    } else {
+      ZeImage = UrImage->ZeImage;
+      *phMem = nullptr;
+    }
+  } else {
+    ze_image_pitched_exp_desc_t PitchedDesc;
+    PitchedDesc.stype = ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC;
+    PitchedDesc.pNext = nullptr;
+    PitchedDesc.ptr = hImageMem;
```

Comment:
is this correct ? 

ptr is expected to be raw pointer returned from  urUSMPitchedAllocExp -> urUSMDeviceAlloc(hContext, hDevice, pUSMDesc, pool, Size, ppMem) ->  ppMem

Reply:
hImageMem can be either an image created by zeImageCreate or usm created by zeMemAllocDevice. So a query is needed at the beginning of this function.
added a check and return error if MemAllocProperties.type isn't device usm

Reply:
ok, 
it wasnt clear to me that  ur_exp_image_mem_handle_t   is interpreted as image handle or just a raw pointer.